src/ompl/base/spaces/special/src/KleinBottleStateSpace.cpp

```diff
@@ -0,0 +1,331 @@
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the MPI-IS nor the names
+ *     of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written
+ *     permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Andreas Orthey */
+
+#include <ompl/base/spaces/special/KleinBottleStateSpace.h>
+#include <ompl/tools/config/MagicConstants.h>
+#include <cstring>
+#include <boost/math/constants/constants.hpp>
+
+using namespace boost::math::double_constants;  // pi
+using namespace ompl::base;
+
+KleinBottleStateSampler::KleinBottleStateSampler(const StateSpace *space) : StateSampler(space)
+{
+}
+
+void KleinBottleStateSampler::sampleUniform(State *state)
+{
+    bool acceptedSampleFound = false;
+    while (!acceptedSampleFound)
+    {
+        double u = rng_.uniformReal(0, pi);
+        double v = rng_.uniformReal(-pi, pi);
+
+        // NOTE: The idea here is to compute the norm of the gradient at each
+        // point of the surface (i.e. the gradient of the coordinate mapping from
+        //(u,v) to (x,y,z)). To get the norm, we divide by the maximum norm of the
+        // gradient over the whole surface. This gives a number between [0,1]. We
+        // then do rejection sampling, by choosing a random number in [0,1] and
+        // accept if the norm is larger than this random number. Surface elements
+        // with a high curvature will have a small norm and will therefore be
+        // penalized under this method (i.e. rejected more often).
+        // See also: https://mathematica.stackexchange.com/questions/148693/generating-random-points-on-a-kleins-bottle
```

Comment:
Yes you are right, points which have a large gradient will have a much smaller surface area, so they will be rarely picked. If they appear, we try to accept them often. If we encounter a small gradient, then we will have a larger surface area and we should sometimes reject some of those samples. This was described for the Torus in the following post on stackoverflow: https://stackoverflow.com/a/55699723/1067803
There is also an excellent paper on the topic: J.F. Williamson, "Random selection of points distributed on curved surfaces", 1987

For the Kleinbottle, I basically just followed this recipe. The only difficulty was that I did not know the maximum stretch factor, and I did some upper-bound estimation by sampling 10k points and observed the max factor.

Reply:
Nice, thanks for taking the time to explain and providing the reference. That's a neat technique to sample uniform points on complex surfaces.

Sampling 10k points seems reasonable to me, so not suggesting you do the following instead. Just curious whether you think it would work. The norm of the gradient is differentiable with respect to u and v, right? So we should be able to find all local maxima  of the norm and compare those to find the global maximum (by "we" I mean sympy or equivalent)? Again, not suggesting you do this, just wondering whether it would work.

Reply:
Seems like a great idea. You mean we randomly sample a point and do gradient descent on it? I think this should work, albeit there might be some convergence errors (so we might be slightly off from the local maximum), but especially for 2D examples, that should work. But yeah, I think for now it should be fine like that.